hadoop-hdds/docs/content/design/s3_hcfs.md

```diff
@@ -0,0 +1,280 @@
+| `s3` key `/a/b/c` available from `ofs/o3fs` | YES | NO
+| `s3` key `/a/b//c` available from `ofs/o3fs` | YES | NO
+| `s3` key `/a/b//c` available from `s3` | AWS S3 incompatibility | YES
+
+This proposal suggest to use a 3rd option where 100% AWS compatiblity is guaranteed in exchange of a limited `ofs/o3fs` view:
+
+| ozone.om.intermediate.dir.generation= | true |
+|-|-|-|
+| create itermediate dirs | YES | 
+| normalize key names from `ofs/o3fs` | YES |
+| force to normalize key names of `s3` interface | **NO** |
+| `s3` key `/a/b/c` available from `ofs/o3fs` | YES | 
+| `s3` key `/a/b//c` available from `ofs/o3fs` | NO | 
+| `s3` key `/a/b//c` available from `s3` | *YES* (100% AWS compatibility) |
+
+
+# Proposed solution
+
+In short: 
+
+ **I propose to make it possible to configure **normalization** and **intermediate dir creation**, independent from each other**
+ 
+It can be done in multiple ways. For the sake of simplicity, let's imagine two configuration option
+
+| configuration | behavior | 
+|-|-|
+| `ozone.om.enable.filesystem.paths=true`  | Enable intermediate dir generation **AND** key name normalization 
+| `ozone.om.enable.intermediate.dirs=true` | Enable only the intermediate dir generation
+
+## S3/HCFS Interoperability
+
+**In case of intermediate directory generation is enabled (with either of the configuraiton keys)**:
+
+When somebody creates a new key like `/a/b/c/d`, the same key should be visible from HCFS (`o3fs//` or `o3://`). `/a`, `/a/b` and `/a/b/c` should be visible as directories from HCFS.
+
+S3 should list only the `/a/b/c/d` keys, (`/a`, `/a/b`, `/a/b/c` keys, created to help HCFS, **won't be visible** if the key is created from S3)
+
+This can be done with persisting an extra flag with the implicit directory entries. These entries can be modified if they are explicit created.
+
+This flag should be added only for the keys which are created by S3. `ofs://` and `of3fs://`  create explicit directories all the time.
+
+Advantages of this approach:
+
+ 1. HCFS and S3 can work together
+ 2. S3 behavior is closer to the original AWS s3 behavior (when `/a/b/c` key is created `/a/b` won't be visible)
+
+## Handling of the incompatible paths
+
+As it's defined above the intermediate directory generation and normalization are two independent settings. (It's possible to choose only to create the intermediate directories).
+
+**If normalization is choosen**: (`ozone.om.enable.filesystem.paths=true`), all the key names will be normalized to fs-compatible name. It may cause a conflict (error) if the normalized key is already exists (or exists as a file instead of directory)
+
+**Without normalization (`ozone.om.enable.intermediate.dirs=true`)**:
+
+Creating intermediate directories might not be possible if path contains illegal characters or can't be parsed as a file system path. **These keys will be invisible from HCFS** by default. They will be ignored during the normal file list.
```

Comment:
The illegal char from S3 path can be encoded into FS path except the /.

Reply:
`..` also can be tricky.

But in general, I agree. We might be more permissive and show some of these elements even in `ofs/o3fs`. For example `a/b/c//d` might be possible to show as `/a/b/c/d`. But even if some of these are visible, we should accept that `ofs/o3fs` doesn't provide a *full* view when 100% AWS compatibility is requested.